### DISCUSSION TITLE
Skip "assign_gRNA" step in local laptop

### DISCUSSION THREAD

**Discussion by nhu-github** [Katsevich-Lab/sceptre#117] [Q&A] [answered]
Hi, as I went through of "assign_gRNA" step, I did not find any parameter setup to skip it even thought I noticed it is an optional step. This is in local laptop. Any suggestion on how to skip this step as my data already assigned gRNAs. https://timothy-barry.github.io/sceptre-book/assign-grnas.html

This try does not work though:
sceptre_object@last_function_called <- as.character('assign_grnas')

```Error in perform_status_check_and_update(sceptre_object, "run_qc") :  The function `assign_grnas()` must be called before `run_qc()` is called ```

**Comment by timothy-barry** [comment]
Hi Nan,

"Optional" in this context means that `sceptre` automatically calls `assign_grnas()` (with default parameters) if the user does not call this function.

Would you mind sharing a bit more detail about your analysis? In particular, did you import the data via `import_data()`? If so, what did you supply to the `grna_matrix` argument? A binary matrix of gRNA-to-cell assignments? If so, my recommendation would be to explicitly call `assign_grnas()`, setting `method` to `"thresholding"` and `threshold` to `1`. This will cause `sceptre` to use your custom assignments.

**Reply by nhu-github** [reply]
Thanks for your reply. So I follow the standard pipeline from the beginning to the end of discovery analysis and save the final sceptre object. This time I re-run `set-analysis-parameter()` to change the `discovery_pairs` to the previous sceptre object which means the gRNA is already assigned in previous analysis. Then I would like to skip `assign_gRNA()`, but the program did not allow me to. `thresholding=1` will change the original gRNA assignment.

**Comment by timothy-barry** [comment, marked as answer]
Oh, I see. So this is a question about caching/optimization. `sceptre` currently supports basic caching, but it is not super sophisticated. We plan to add more sophisticated caching functionality in a future update.

For the time being, I would recommend calling `sceptre_object <- sceptre:::perform_status_check_and_update(sceptre_object, "assign_grnas")` *instead of* `assign_grnas()`, as follows.

```
library(sceptredata)
library(sceptre)

directories <- paste0(
  system.file("extdata", package = "sceptredata"),
  "/highmoi_example/gem_group_", c(1, 2)
)
data(grna_target_data_frame_highmoi)
sceptre_object <- import_data_from_cellranger(
  directories = directories,
  moi = "high",
  grna_target_data_frame = grna_target_data_frame_highmoi
)
print(sceptre_object)

# 2. set the analysis parameters
positive_control_pairs <- construct_positive_control_pairs(sceptre_object)
discovery_pairs <- construct_cis_pairs(sceptre_object,
                                       positive_control_pairs = positive_control_pairs,
                                       distance_threshold = 5e6
)

sceptre_object <- set_analysis_parameters(
  sceptre_object = sceptre_object,
  discovery_pairs = discovery_pairs,
  positive_control_pairs = positive_control_pairs,
  side = "left"
)
print(sceptre_object)

# 3. assign grnas
plot_grna_count_distributions(sceptre_object)
sceptre_object <- sceptre_object |> assign_grnas(parallel = TRUE, n_processors = 2)
plot(sceptre_object)
print(sceptre_object)

# 4. run qc
plot_covariates(sceptre_object, p_mito_threshold = 0.075)
sceptre_object <- sceptre_object |> run_qc(p_mito_threshold = 0.075)
plot(sceptre_object)
print(sceptre_object)

# 5. run the calibration check
sceptre_object <- run_calibration_check(sceptre_object, parallel = TRUE, n_processors = 2)
plot(sceptre_object)
print(sceptre_object)

# 6. run the power check
sceptre_object <- run_power_check(sceptre_object, parallel = TRUE, n_processors = 2)
plot(sceptre_object)
print(sceptre_object)

# 7. run discovery analysis
sceptre_object <- run_discovery_analysis(sceptre_object, parallel = TRUE, n_processors = 2)
plot(sceptre_object)
print(sceptre_object)

# 8. reassign the discovery pairs
discovery_pairs <- construct_cis_pairs(sceptre_object,
                                       positive_control_pairs = positive_control_pairs,
                                       distance_threshold = 1e6
)
sceptre_object <- set_analysis_parameters(
  sceptre_object = sceptre_object,
  discovery_pairs = discovery_pairs,
  positive_control_pairs = positive_control_pairs,
  side = "left"
)

# 9. skip the grna assignment step
sceptre_object <- sceptre:::perform_status_check_and_update(sceptre_object, "assign_grnas")

# 10. proceed with the rest of  theanalysis
sceptre_object <- sceptre_object |> run_qc() 
sceptre_object <- run_calibration_check(sceptre_object, parallel = TRUE, n_processors = 2) # etc
```

**Reply by nhu-github** [reply]
Thanks. This works.
